### DISCUSSION TITLE
How to Remove Checkbox Button Widget

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [pyvista/pyvista#3303] [Q&A] [answered]
Borrowing from the [Single Checkbox Widget Example](https://docs.pyvista.org/examples/03-widgets/checkbox-widget.html#single-checkbox), what is the correct way to remove a checkbox button from a plotter?

e.g. If I try to manually delete the button, I find it is still around and active. Is this because the `plotter` still might hold a reference to it?

I can "cheat" by hiding the button if I set its opacity to 0 and turn it off:

```python
button.GetRepresentation().GetProperty().SetOpacity(0.0)
button.Off()
```

but I would really like to know how to remove the object so I can free memory when I need to. There seem to be methods for removing other widgets in the API, but they seem to be little more than clearing a list. I would have assumed if those worked that below would also work, but that appears not to be the case. Perhaps I'm missing something?

### MRE

```python
import pyvista as pv

mesh = pv.Sphere()

p = pv.Plotter()
actor = p.add_mesh(mesh)


def toggle_vis(flag):
    actor.SetVisibility(flag)


button = p.add_checkbox_button_widget(toggle_vis, value=True)

button = None  # This does not remove the button...
del button  # ...nor does this

p.show()
```

![Lingering_Checkbox_Widget](https://user-images.githubusercontent.com/59346180/189791267-d25f6d8b-4e05-4947-9646-4cb9ae7cd66f.gif)

### Update

I've also tried using

```python
p.clear_button_widgets()
```

but that also does not seem to work...

**Comment by adam-grant-hendry** [comment, marked as answer]
This was a "dumb" mistake on my end. In the above, **both** `p.clear_button_widgets()` and `del button` are required:

- [WidgetHelper](https://github.com/pyvista/pyvista/blob/b0143f3bc3fd2c0d5b7cef6d7f7838b8cf07fc20/pyvista/plotting/widgets.py#L18) holds references to all widgets as lists. [BasePlotter](https://github.com/pyvista/pyvista/blob/b0143f3bc3fd2c0d5b7cef6d7f7838b8cf07fc20/pyvista/plotting/plotting.py#L132) inherits `WidgetHelper` and `PickingHelper` and [Plotter](https://github.com/pyvista/pyvista/blob/b0143f3bc3fd2c0d5b7cef6d7f7838b8cf07fc20/pyvista/plotting/plotting.py#L5594) inherits `BasePlotter`
- `button` is not a `weakref`, so it increases the refcount by 1

To clear the button, there are 3 options:

### 1. Clear the Button Widgets List and Delete Button

```python
p.clear_button_widgets()

button = None
del button
```

### 2. Don't Keep a Reference to Button

```python
...
p.add_checkbox_button_widget(toggle_vis, value=True)
...
p.clear_button_widgets()
```

### 3. Make `button` a `weakref`

```python
import weakref
import pyvista as pv
...
# NOTE: ``weakref.ref``s are callables. To access the underlying reference, you would
# write ``button()``. If the object has been deleted, this call returns ``None``
button = weakref.ref(p.add_checkbox_button_widget(toggle_vis, value=True))
...
p.clear_button_widgets()
```

@akaszynski This might be a good thing to add as a `NOTE`/`WARNING`/`GOTCHA` to the documentation. It might bite people unexpectedly.

**Reply by banesullivan** [reply]
great catch, @adam-grant-hendry! We may want to change what we can to be a `weakref` or add a note in the documentation as you outline

**Reply by adam-grant-hendry** [reply]
@banesullivan Followup question: If I want the button to do something, then go away when it's done, how can I do that?

1. Trying below, the hand cursor lingers and the button doesn't disappear until the next click

```python
import pyvista as pv

mesh = pv.Sphere()

p = pv.Plotter()
actor = p.add_mesh(mesh)


def toggle_vis(flag):
    # Do something
    p.clear_button_widgets()


p.add_checkbox_button_widget(toggle_vis, value=True)

p.show()
```

https://user-images.githubusercontent.com/59346180/189959190-6a5e8dd3-e6ea-4352-93f1-4e327245edf4.mp4

I see `checkbox_button`s respond to `vtkCommand.StateChangedEvent`. Do I need to handle `vtkCommand.EndInteractionEvent` somehow?

**Reply by adam-grant-hendry** [reply]
@banesullivan @akaszynski I moved this last question into #3316 , since I'm not sure if this is a bug or not (I might be doing another thing wrong).
